lib/history-index.js

```diff
@@ -0,0 +1,110 @@
+var duplexify = require('duplexify')
+var through = require('through2')
+
+function createIndex (osm, db) {
+  return {
+    maxBatch: 100,
+    map: function (nodes, next) {
+      var batch = []
+      nodes.forEach(function (node) {
+        if (node.value.type !== 'osm/element') return
+        if (!node.value.element) return
+        var key = node.value.element.type === 'observation'
+          ? 'created_at' : 'timestamp'
+        var timestamp = node.value.element[key]
+        if (!timestamp) return
+        // dates should be stored as ISO format which lexicographically sort
+        // but we'll convert and re-encode just to be extra sure
+        var date = new Date(timestamp)
+        try {
+          var iso = date.toISOString()
+        } catch (err) {
+          return
+        }
+        batch.push({
+          type: 'put',
+          key: 'a!' + iso,
+          value: node.key + '@' + node.seq
+        })
+        var t = String(node.value.element.type || '').replace(/!/g, '_')
+        batch.push({
+          type: 'put',
+          key: 't!' + t + '!' + iso,
+          value: node.key + '@' + node.seq
+        })
+      })
+      if (batch.length === 0) next()
+      else db.batch(batch, next)
+    },
+    storeState: function (state, cb) {
+      db.put('_state', state, cb)
+    },
+    fetchState: function (cb) {
+      db.get('_state', function (err, state) {
+        if (err && err.notFound) cb()
+        else if (err) cb(err)
+        else cb(null, state)
+      })
+    },
+    api: {
+      type: function (core, type, opts) {
+        var d = duplexify.obj()
+        this.ready(function () {
+          var s = through.obj(write)
+          var r = db.createReadStream(prefixOpts('t!' + type, opts))
+          r.pipe(s)
+          r.on('error', function (err) { s.emit('error', err) })
+          d.setReadable(s)
+        })
+        return d
+        function write (row, enc, next) {
+          osm.getByVersion(row.value, function (err, doc) {
+            if (err) next(err)
```

Comment:
It shouldn't happen generally, but `getByVersion` can return `null` docs and no error, in the case that you somehow replicated references to messages but not those messages yet. Could we do a `if (!doc) return next(null)` after the error check to avoid nulls in the results?

Reply:
^ same in the `all` api below

Reply:
added this case to both `type` and `all` methods